server/master/src/main/java/org/apache/accumulo/master/TimeoutTaskExecutor.java

```diff
@@ -0,0 +1,235 @@
+
+    this.executorService = Executors.newFixedThreadPool(numThreads);
+    this.startedTasks = new ArrayBlockingQueue<>(expectedNumCallables);
+    this.timeout = timeout;
+    this.wrappedTasks = new ArrayList<>(expectedNumCallables);
+  }
+
+  /**
+   * Submits a new task to the executor.
+   *
+   * @param callable Task to run
+   */
+  public void submit(C callable) {
+    WrappedTask wt = new WrappedTask(callable);
+    wt.future = executorService.submit(wt);
+    wrappedTasks.add(wt);
+  }
+
+  /**
+   * Registers the callback to use on successful tasks.
+   *
+   * @param successCallback The callback function to invoke on success.
+   * @throws NullPointerException when a null successCallback is passed in
+   */
+  public void onSuccess(SuccessCallback<T,C> successCallback) {
+    this.successCallback = Objects.requireNonNull(successCallback, "Must provide a non-null successCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that throw exceptions.
+   *
+   * @param exceptionCallback The callback function to invoke on exceptions.
+   * @throws NullPointerException when a null exceptionCallback is passed in
+   */
+  public void onException(ExceptionCallback<C> exceptionCallback) {
+    this.exceptionCallback = Objects.requireNonNull(exceptionCallback, "Must provide a non-null exceptionCallback.");
+  }
+
+  /**
+   * Registers the callback to use on tasks that time out.
+   *
+   * @param timeoutCallback The callback function to invoke on timeouts.
+   * @throws NullPointerException when a null timeoutCallback is passed in
+   */
+  public void onTimeout(TimeoutCallback<C> timeoutCallback) {
+    this.timeoutCallback = Objects.requireNonNull(timeoutCallback, "Must provide a non-null timeoutCallback.");
+  }
+
+  /**
+   * Completes all the current tasks by dispatching to the appropriate callback.
+   *
+   * @throws IllegalStateException If all of the callbacks were not registered before calling this method.
+   * @throws InterruptedException  If interrupted while awaiting callable results.
+   */
+  public void complete() throws InterruptedException {
+    Preconditions.checkState(successCallback != null, "Must set a success callback before completing " + this);
+    Preconditions.checkState(exceptionCallback != null, "Must set an exception callback before completing " + this);
+    Preconditions.checkState(timeoutCallback != null, "Must set a timeout callback before completing " + this);
+
+    while (hasUnfinishedTasks()) {
+      completeTask(startedTasks.take());
+    }
+
+    wrappedTasks.clear();
+  }
+
+  private boolean hasUnfinishedTasks() {
+    for (WrappedTask wt : wrappedTasks) {
+      if (!wt.hasCompleted) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  private void completeTask(WrappedTask wt) throws InterruptedException {
+    try {
+      long waitTime = wt.endTime - System.currentTimeMillis();
```

Comment:
For sanity, I think it would be nice to have a getEndTime() method that ensures that endTime was set. Something like

```java
long getEndTime(){
  long et = endTime; //only read volatile once
  Preconditions.checkState(et > 0);
  return et;
}
```

Its possible that System.currentTimeMillis() can go backwards which would result in a negative `waitTime`.  I would normally recommend using System.nanotime() to avoid this.  However I was reading that nanoTime may only be good for use within the same thread/cpu, but I think this was only true on older OS.  This code would call nanotime in diff threads.  I think using nanoTime and/or handling negative waitTime would be good.  If its negative could just assume a waitTime of zero.

Reply:
This hasn't been an issue on Linux since very old Linux 2.6 kernels prior to 2.6.18, which itself hasn't shipped since RHEL5, and even then, the problem occurred only on some hardware. The java source code uses `CLOCK_MONOTONIC` if available, and the same source as `System.currentTimeMillis()` if not available (to millisecond resolution in the latter case).

Absolute times can be negative, but elapsed time / durations / time spans, should always be positive. If not, it's an OS bug. I seriously doubt this is something we need to worry about on modern Linux or Mac/BSD systems.

I seem to recall debating this when similar reservations were raised a few years ago when I suggested using nanoTime more back then. It wasn't a current know bug then on any systems we support, and it isn't now. I think it's fine to use it.

Reply:
Will update to nanoTime(). Thanks for the info guys. Learned something new :)